### ISSUE TITLE
colormaps (cm) do not properly handle NaN values.

### ISSUE THREAD

**Issue by amicitas** [matplotlib/matplotlib#9892]
### Bug report

**Bug summary**

When using a colormap to generate a color array with data that contains np.nan (NaN), an warning is given, and the previous color entry is duplicated.  The method that has this issue is `to_rgba`.

**Code for reproduction**

```python
import numpy as np
from matplotlib.colors import Normalize
from matplotlib import cm

y = np.linspace(0.0, 1.0, 4)
y[1] = np.nan

norm = Normalize(0.0, 1.0)
colormap = cm.ScalarMappable(norm, 'rainbow')
colors = colormap.to_rgba(y)

print(colors)
```

**Actual outcome**

```
/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/matplotlib/colors.py:496: RuntimeWarning: invalid value encountered in less
  cbook._putmask(xa, xa < 0.0, -1)

[[  5.00000000e-01   0.00000000e+00   1.00000000e+00   1.00000000e+00]
 [  5.00000000e-01   0.00000000e+00   1.00000000e+00   1.00000000e+00]
 [  8.33333333e-01   8.66025404e-01   5.00000000e-01   1.00000000e+00]
 [  1.00000000e+00   1.22464680e-16   6.12323400e-17   1.00000000e+00]]
```

**Expected outcome**

NaN values should be quietly handled.  I think the most reasonable output would be to fill the color array with np.nan rather than copy the previous entry.

```
[[  5.00000000e-01   0.00000000e+00   1.00000000e+00   1.00000000e+00]
 [  NaN              NaN              NaN              NaN]
 [  8.33333333e-01   8.66025404e-01   5.00000000e-01   1.00000000e+00]
 [  1.00000000e+00   1.22464680e-16   6.12323400e-17   1.00000000e+00]]
```

**Matplotlib version**

  * Operating system: OSX
  * Matplotlib version: 2.0.2
  * Matplotlib backend: module://ipykernel.pylab.backend_inline & MacOSX
  * Python version: 3.6.3
  * Jupyter Notebook version: 5.2.0
  * Other libraries: None

**Comment by efiring** [comment]
You are making an assumption about how NaNs should be handled that is inconsistent with Matplotlib's design.  Originally, we used masked arrays exclusively to handle missing values.  As support for NaNs improved in the underlying numerical packages (numpy replaced Numeric and numarray), we added calls to `np.ma.masked_invalid` to convert user input with NaNs to masked arrays.  These calls were added to the highest-level parts of the API, like `plot`; they are not repeated at every stage below that.  The `to_rgba` method is an example of a slightly lower level of the API; it is called within higher-level functions and methods, and is not normally needed at the outermost "user" level.  Therefore it does not support NaNs in its input. Furthermore, NaNs are not supported in any color specification.  NaNs or masked values in user-level input are taken to mean "don't plot this" in routines like `plot` or "use a color specifically designated to mean this is invalid" in routines like `contourf` and `pcolormesh`, and are then handled appropriately at the lower levels.
Therefore `to_rgba` always returns a valid color specification.

**Comment by tacaswell** [comment]
You can get this more directly:

```python
import numpy as np
from matplotlib.colors import Normalize
from matplotlib import cm

y = np.linspace(0.0, 1.0, 4)
y[1] = np.nan

norm = Normalize(0.0, 1.0)
v = cm.get_colormap()
print(v(norm(y)))

```

If you mask the invalid entries you get:

```python
In [34]: v(np.ma.masked_invalid(norm(y)))
Out[34]: 
array([[ 0.267004,  0.004874,  0.329415,  1.      ],
       [ 0.      ,  0.      ,  0.      ,  0.      ],
       [ 0.20803 ,  0.718701,  0.472873,  1.      ],
       [ 0.993248,  0.906157,  0.143936,  1.      ]])

```

which maps `np.nan` to the 'bad' color (which is by default (0, 0, 0, 0)).  

Returning `nan` is untenable, the output from this function is passed to the various backends as RGBA colors so it _must_ be a float in the range [0, 1].

The reason you are seeing this behavior is some fun numpy-related casting issues:

```python
import numpy as np
print()
# two good entries in LUT
N = 2

i_under = 2
i_over = 3
i_bad = 5

# color Look Up Table as strings to be readable
lut = np.array(
    ['v0',
     'v1',
     'under',
     'over',
     'bad'])

# the input
f = np.array([0, np.nan, .7, 2])
print(f'input: \n\t{f}')
# scaled to map to lookup table
f *= N
print(f'scaled to LUT size \n\t{f}')
# converted to ints (which we need
# for the lookup table that is coming
xa = f.astype(int)
print(f'mapped to int \n\t{xa}')
# map 'over' values to the over index
np.copyto(xa, i_over, where=(xa > N - 1))
print(f'over values mapped \n\t{xa}')
# map 'under' values to the under index
np.copyto(xa, i_under, where=(xa < 0))
print(f'under values mapped \n\t{xa}')
mapped = lut.take(xa, axis=0, mode='clip')
print(f'mapped to lut \n\t{mapped}')

```

which gives

```
input: 
	[ 0.   nan  0.7  2. ]
scaled to LUT size 
	[ 0.   nan  1.4  4. ]
mapped to int 
	[                   0 -9223372036854775808                    1
                    4]
over values mapped 
	[                   0 -9223372036854775808                    1
                    3]
under values mapped 
	[0 2 1 3]
mapped to lut 
	['v0' 'under' 'v1' 'over']

```

The reason that it looks like it is repeating the previous entry is that the default 'under' value is equal to the bottom of the color map and the first entry happens to be 0, for example

```
In [72]: v(norm([.5, np.nan, .1]))
/home/tcaswell/src/p/matplotlib/lib/matplotlib/colors.py:475: RuntimeWarning: invalid value encountered in less
  np.copyto(xa, -1, where=xa < 0.0)
Out[72]: 
array([[ 0.127568,  0.566949,  0.550556,  1.      ],
       [ 0.267004,  0.004874,  0.329415,  1.      ],
       [ 0.282623,  0.140926,  0.457517,  1.      ]])
```

The actually implementation of the LUT that I cribbed the above from:

https://github.com/matplotlib/matplotlib/blob/d046efbc210696f7041b2e97e9f898a2a09bf59a/lib/matplotlib/colors.py#L462-L507

**Comment by tacaswell** [comment]
We should probably distill some combination of my and @efiring 's comments into the color map `__call__` docstring and maybe the `to_rgba` docstring.

**Comment by amicitas** [comment]
Thank you @efiring for the detailed explanation.  This is very useful to have documented, even as part of a bug report. I understand that having `to_rgba` give a valid color specification makes sense (and is required) in the context of how matplotlib backend handles colors.

If any points with NaN were mapped consistently to the 'bad color', then the output would be immediately sensible and understandable.  This would also work reasonably well for visualizing data that contains NaN. It would work even better if `to_rgba` would accept a keyword that would set the value of the 'bad color'.

To get around the numpy casting issues, is it possible to have `np.ma.masked_invalid` always automatically applied to the input in colormaps?

I do think it is important that colormap and norm have reasonable handling of NaN. In certain types of scientific analysis NaN is a valid and meaningful floating point value that needs to be part of data sets.

**Comment by greglucas** [comment]
I've run into this issue several times where I forget to mask the array and want the nan's to get the 'bad' value, but they end up getting the 'under' value instead. The -inf and +inf get mapped properly to under/over respectively, due to the `np.clip` call, so the only thing ambiguous is nan I believe.

https://github.com/matplotlib/matplotlib/blob/d0628598f8d9ec7b0da6b60e7b29be2067b6ea17/lib/matplotlib/colors.py#L481-L494

Adding an extra check for nans immediately after the over/under check would fix the nan to under conversion.
`xa[np.isnan(X)] = self._i_bad`

Simple example:
```python
import numpy as np
import matplotlib.pyplot as plt
x = np.array([-np.inf, -1, 0, np.nan, .7, 2, np.inf])
cmap = plt.cm.plasma
cmap(x)
```

Current output (with nan cast to 'under'):
```
array([[0.050383, 0.029803, 0.527975, 1.      ],
       [0.050383, 0.029803, 0.527975, 1.      ],
       [0.050383, 0.029803, 0.527975, 1.      ],
       [0.050383, 0.029803, 0.527975, 1.      ],
       [0.949217, 0.517763, 0.295662, 1.      ],
       [0.940015, 0.975158, 0.131326, 1.      ],
       [0.940015, 0.975158, 0.131326, 1.      ]])
```

New output with the change I mentioned (what I expect):
```
array([[0.050383, 0.029803, 0.527975, 1.      ],
       [0.050383, 0.029803, 0.527975, 1.      ],
       [0.050383, 0.029803, 0.527975, 1.      ],
       [0.      , 0.      , 0.      , 0.      ],
       [0.949217, 0.517763, 0.295662, 1.      ],
       [0.940015, 0.975158, 0.131326, 1.      ],
       [0.940015, 0.975158, 0.131326, 1.      ]])
```

I'm happy to make a PR with this change if you think this update is acceptable.

**Comment by tacaswell** [comment]
I am pretty sure we do not currently do that for performance reasons.

**Comment by jklymak** [comment]
Hard to see how that is any more expensive than all the other manipulations we are doing here.

**Comment by greglucas** [comment]
I just ran a few %%timeit tests and it is honestly mostly in the noise and hard to tell a difference in performance. Maybe a few percent in time and the memory hit is basically nothing as well as we aren't saving the nan-tested array.

There are a few other ways to go about this as well because the only thing outside of the clip range would be the really small int from the nan conversion.

- Check if the int conversion created any minimum integers.
`xa[xa == np.iinfo(xa.dtype).min] = self._i_bad`
Note that this would change some behavior if someone was already passing in an integer array that had the most negative value already (although probably for the better to warn of bad values from previous inf/-inf conversions).

**Comment by tacaswell** [comment]
I am convinced.

@greglucas Would you mind putting that change (and that example as a test) into a PR?

**Comment by greglucas** [comment]
Closed by #14257.

**Comment by QuLogic** [comment]
FYI, if you changed how you [referenced this issue in the PR slightly](https://help.github.com/en/articles/closing-issues-using-keywords), it will close automatically.
